16494 Protractor (JavaScript e2e) tests

myabc opened the pull request:
:warning: **Includes commits from #2067. Please review, merge first.**

https://www.openproject.org/work_packages/16494

myabc committed:
Add Protractor config, example e2e spec

This initial spec still relies on a running Rails server (with default
seed data).

User Story # 16494

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Add protractor, chai-as-promised dependencies

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Move from grunt to gulp build system

Initially port only JSHint tasks.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Add basic JSHint config

Tolerate subscript notation in addition to dot notation.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Document Protractor in RUNNING_TESTS

Signed-off-by: Alex Coles <alex@alexbcoles.com>

TeatroIO commented:
I've prepared a stage. <a href='http://feature-16494-protractor-tests.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Click to open.</a>

myabc committed:
Bundle CommonJS modules with Webpack

Minimal work necessary to make Sprockets and existing Karma tests work
with Webpack.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Move angular app requires to openproject-app.js

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Merge branch 'fix/normalise-file-naming' into feature/16494-protractor-tests

myabc committed:
Restructure Angular app into CommonJS modules

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Fix stray global in Karma tests

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Merge branch 'dev' into feature/16494-protractor-tests

Signed-off-by: Alex Coles <alex@alexbcoles.com>

Conflicts:
	app/assets/javascripts/angular/work_packages/view_models/common-relations-handler.js

myabc commented on app/assets/javascripts/angular/work_packages/tabs/index.js:
```diff
@@ -0,0 +1,70 @@
+    require('./work-package-relations-directive')
+  ]);
+
+// FIXME: move modules or files to the right place
```
@hschink any idea why these directives were placed in a different module?

hschink replied:
Sorry, I have no idea

myabc committed:
Add Rake tasks for Webpack compilation

Hook Webpack build into standard Sprockets precompilation.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Hook Webpack build into test/spec tasks

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Only hook into test/spec tasks if they are defined

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Configure Travis to do always npm install

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Remove duplicate bower steps in Travis config

As of 631147af, `npm install` should automatically invoke a
`bower install`.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Configure Travis to also cache npm dependencies

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Hook Webpack build into Cucumber tasks

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
First steps to bootstrap Angular app without Rails

Create basic express app that serves a static HTML page for
bootstrapping the Angular app and a minimal subset of API endpoints.

* Create Gulp tasks for starting express server, compiling Sass.

User Story # 16494

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Remove usages of Prototype-added Array methods

Replace with equivalent lodash methods.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Add gulp-protractor, gulp tasks for Protractor

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Remove login steps from Protractor spec

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Bundle I18n translations with Webpack

Vendor i18n.js script provided by fnando/i18n-js.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Remove seleniumAddress from Protractor config

This should ensure Protractor spins up its own bundled Selenium
server.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Fix expected wp columns in example e2e spec

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Merge branch 'dev' into feature/16494-protractor-tests

Signed-off-by: Alex Coles <alex@alexbcoles.com>

Conflicts:
	app/assets/javascripts/angular/work_packages/view_models/common-relations-handler.js

myabc committed:
Add npm scripts entry for Protractor

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Configure Travis to execute Protractor tests

User Story # 16494 (Task # 16553)

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Use Firefox (Travis-supported) with Protractor

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Compile Sass before executing Protractor specs

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Normalise naming of more Sass partials

See 457b1496.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Bring in various styles with require directive

We reserve `@import` for pure-Sass (non-erb) partials. See 240c7eeb.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Use default.css stylesheet for Protractor, LSG

As Sprockets directives are ignored by the Gulp sass plugin, b6d5c298
should obviate the need for a separate 'meta' stylesheet.

This partially reverts 271aeef434ebb31e57284606a87fced974768e16.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Move .nomenus padding rules out of legacy styles

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc commented on .travis.yml:
```diff
@@ -30,7 +30,9 @@ language: ruby
-cache: bundler
+cache:
+  - bundler
+  - npm
```
N.B. I don't believe this is working yet.

myabc committed:
Add back inadvertently deleted @import

See b6d5c298.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Restructure RUNNING_TESTS, update Protractor info

[ci skip]

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Show Cucumber, TU as deprecated in RUNNING_TESTS

[ci skip]

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Remove unnecessary steps from INSTALL

As of 631147af, `npm install` should automatically invoke a
`bower install`

[ci skip]

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Add Webpack step to INSTALL doc

[ci skip]

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Add missing Copyright headers to mock API

[ci skip]

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Remove hardcoded host from Protractor specs

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc commented:
When reviewing, please bear in mind updated INSTALL and RUNNING_TESTS docs (see d1b778d and da316b7).

myabc committed:
Change Express server port to 8080

This should make conflicts with a running Rails server less likely.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc commented:
@0xF013 please take a look!

myabc committed:
Rework example e2e spec to use Page Object pattern

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Merge branch 'dev' into feature/16494-protractor-tests

Signed-off-by: Alex Coles <alex@alexbcoles.com>

Conflicts:
	package.json

0xF013 commented:
@myabc I am having some troubles starting protractor

```
$ gulp tests:protractor
[11:26:36] Using gulpfile ~/Work/openproject/gulpfile.js
[11:26:36] Starting 'webdriver:update'...
[11:26:36] Starting 'webpack'...
[11:26:37] Starting 'sass'...
[11:26:37] Starting 'express'...
selenium standalone is up to date.
chromedriver is up to date.
[11:26:37] Finished 'express' after 114 ms
Listening at localhost:8080
[11:26:37] Finished 'webdriver:update' after 154 ms
[11:26:38] gulp-ruby-sass: directory
[11:26:38] Version: webpack 1.4.9
                    Asset     Size  Chunks             Chunk Names
openproject-app.bundle.js  1941589       0  [emitted]  main
[11:26:38] Finished 'webpack' after 1.23 s
[11:26:38] gulp-ruby-sass: write default.css.css
[11:26:38] Finished 'sass' after 1.96 s
[11:26:38] Starting 'tests:protractor'...
Starting selenium standalone server...

/Users/elvis/Work/openproject/node_modules/protractor/node_modules/selenium-webdriver/lib/webdriver/promise.js:1745
      throw error;
            ^
Error: Timed out waiting for the WebDriver server at http://10.1.2.32:51517/wd/hub
    at Error (<anonymous>)
    at onResponse (/Users/elvis/Work/openproject/node_modules/protractor/node_modules/selenium-webdriver/http/util.js:87:11)
    at /Users/elvis/Work/openproject/node_modules/protractor/node_modules/selenium-webdriver/http/util.js:42:21
    at /Users/elvis/Work/openproject/node_modules/protractor/node_modules/selenium-webdriver/lib/webdriver/http/http.js:96:5
    at ClientRequest.<anonymous> (/Users/elvis/Work/openproject/node_modules/protractor/node_modules/selenium-webdriver/http/index.js:145:7)
    at ClientRequest.EventEmitter.emit (events.js:95:17)
    at Socket.socketErrorListener (http.js:1547:9)
    at Socket.EventEmitter.emit (events.js:95:17)
    at net.js:440:14
    at process._tickCallback (node.js:415:13)

/Users/elvis/Work/openproject/gulpfile.js:86
      throw e;
            ^
Error in plugin 'gulp-protractor'
protractor exited with code 8
    at ChildProcess.<anonymous> (/Users/elvis/Work/openproject/node_modules/gulp-protractor/index.js:59:27)
    at ChildProcess.EventEmitter.emit (events.js:98:17)
    at Process.ChildProcess._handle.onexit (child_process.js:797:12)
```

myabc committed:
Fix code indenting in RUNNING_TESTS 

[ci skip]

0xF013 commented:
@myabc will we have code minification?

myabc commented:
@0xF013 yes, but we first have to a) either audit our codebase and make sure we use explicit annotations (Upgrading to Angular 1.3. would [help with this](https://docs.angularjs.org/error/$injector/strictdi)) OR b) incorporate a plugin like ng-annotate to annotate explicit dependencies before minification.
